incubator/vault/templates/deployment.yaml

```diff
@@ -22,21 +28,29 @@ spec:
         {{- if .Values.vault.dev }}
         command: ["vault", "server", "-dev", "-dev-listen-address", "[::]:8200"]
         {{- else }}
-        command: ["vault", "server", "-config", "/vault/config/config.json"]
+        command: ["sh", "-c"]
+        args:
+          - export VAULT_CLUSTER_ADDR="http://$POD_IP:8201"; vault server -config /vault/config/config.json
         {{- end }}
         ports:
         - containerPort: {{ .Values.service.port }}
         livenessProbe:
           tcpSocket:
             port: {{ .Values.service.port }}
         readinessProbe:
-          tcpSocket:
-            port: {{ .Values.service.port }}
+            httpGet:
+              path: /v1/sys/health
```

Comment:
I had to use the following to make it work: `/v1/sys/health?sealedcode=204&standbycode=204&uninitcode=204` (299 would have also worked, but I needed to add the other query parameters)

Reply:
@lvicentesanchez Those other codes, I left it off, cause i don't want the pod health (ready) if it is not unsealed and working.
So i have to internally, unseal/init and etc.. 

From what I understood, you want it to be ready even if it is not unseal. From my perspective, it is not ready yet. 
I might have miss understood what you tried to do. If so, please let me know.

Either way, i think this is more of a project decision.

Maybe be best to template those codes and let it at `values.yml`  I thought no one may want this ready if not unsealed.

Reply:
You are right, I should have unsealed them manually instead of adding those codes.

Reply:
Actually, I was talking about the livenessProbe. For the readinessProbe I agree that it should wait until the vault is initalised and unsealed.

Reply:
So... if I change the readinessProbe, I don't actually need to return custom codes in the livenessProbe

Reply:
liveness probe doesn't need to be httpget, you can use just tcpsocket port to see if the port is listening on, so vault is running. 

But sure, you can have httpget with custom code as well if you want.

Reply:
The purpose of the `readinessProbe` is to ensure requests always go to the leader. See this explanation for why this is especially important in a load balanced scenario (https://www.vaultproject.io/docs/concepts/ha.html#behind-load-balancers):

> However, if the only access to the Vault servers is via the load balancer, the api_addr on each node should be the same: the address of the load balancer. Clients that reach a standby node will be redirected back to the load balancer; at that point hopefully the load balancer's configuration will have been updated to know the address of the current leader. This can cause a redirect loop and as such is not a recommended setup when it can be avoided.

In cases where a client is accessing Vault with a `ClusterIP` or an external load balancer, this is really useful.

Reply:
For some reason I thought the http health point was in the livenessProbe and that's why I made those changes, but they aren't necessary. I tested @sfrique suggestions on top of the changes of this PR and they work flawlessly.

Reply:
@chrismoos Do you intentionally got readiness probe to always fails if it is not the master?
It does work with a load balancer in front of it, I am using the ingress controller, it just forwards the requests internally.

Reply:
Yeah, I did that intentionally. Are you maybe encountering a redirect loop as noted in the Vault docs when your request hits a standby node? Can you try hitting it with `curl -v` a few times and see the redirect behavior?

Reply:
@chrismoos here it follows, i think we can get a lot of information from this i will post it!
I will be using httpie and curl.

```
$ kubectl get pods -o wide | grep vault   
vault-ui-vault-ui-b6fc4ff59-gnhgm                              1/1       Running   0          3d        10.60.2.54   gke-kube-test-1-default-pool-3ded7bfb-z15k
vault-vault-75b9887f95-rnpsj                                   1/1       Running   0          3d        10.60.2.53   gke-kube-test-1-default-pool-3ded7bfb-z15k
vault-vault-75b9887f95-t9xzh                                   1/1       Running   0          5d        10.60.1.46   gke-kube-test-1-default-pool-3ded7bfb-v2vg
vault-vault-75b9887f95-zkr4q                                   1/1       Running   0          5d        10.60.2.52   gke-kube-test-1-default-pool-3ded7bfb-z15k

$ kubectl get service vault-vault
NAME          TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)    AGE
vault-vault   ClusterIP   10.63.244.240   <none>        8200/TCP   5d

$ kubectl describe service vault-vault 
Name:              vault-vault
Namespace:         default
Labels:            app=vault
                   chart=vault-0.2.1
                   heritage=Tiller
                   release=vault
Annotations:       <none>
Selector:          app=vault,release=vault
Type:              ClusterIP
IP:                10.63.244.240
Port:              vault  8200/TCP
TargetPort:        8200/TCP
Endpoints:         10.60.1.46:8200,10.60.2.52:8200,10.60.2.53:8200
Session Affinity:  None
Events:            <none>
```

Just to show curl and httpie behavior. (i've called with curl a few times and always get 200 with no redirects.)

```
$ curl -k -v -H "X-Vault-Token: X" "https://vault-serverv.hiden/secret/qa/test" 
*   Trying X...
* TCP_NODELAY set
* Connected to vault.server.hiden (X) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=vault.hiden
*  start date: Dec  5 14:03:45 2017 GMT
*  expire date: Mar  5 14:03:45 2018 GMT
*  issuer: CN=Fake LE Intermediate X1
*  SSL certificate verify result: unable to get local issuer certificate (20), continuing anyway.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x555de2e70160)
> GET /v1/secret/qa/test HTTP/2
> Host: vault.hiden
> User-Agent: curl/7.57.0
> Accept: */*
> X-Vault-Token: X
> 
* Connection state changed (MAX_CONCURRENT_STREAMS updated)!
< HTTP/2 200 
< server: nginx/1.13.5
< date: Mon, 11 Dec 2017 19:37:51 GMT
< content-type: application/json
< content-length: 198
< cache-control: no-store
< strict-transport-security: max-age=15724800; includeSubDomains;
< 
{"request_id":"a8f7445e-e6e8-d06d-037f-79784a23f696","lease_id":"","renewable":false,"lease_duration":2764800,"data":{"pass":"testpass","user":"teste"},"wrap_info":null,"warnings":null,"auth":null}
* Connection #0 to host vault.hiden left intact
```

No with httpie:
```
$ http --verify=no "https://vault.hiden/v1/secret/qa/test" "X-Vault-Token: X"
HTTP/1.1 200 OK
Cache-Control: no-store
Connection: keep-alive
Content-Length: 198
Content-Type: application/json
Date: Mon, 11 Dec 2017 19:40:07 GMT
Server: nginx/1.13.5
Strict-Transport-Security: max-age=15724800; includeSubDomains;

{
    "auth": null,
    "data": {
        "pass": "testpass",
        "user": "teste"
    },
    "lease_duration": 2764800,
    "lease_id": "",
    "renewable": false,
    "request_id": "4cc66301-a61e-cd9a-9b02-e7d78263f373",
    "warnings": null,
    "wrap_info": null
}
```

From now on, i will be using httpie for all request.

3 request to internal service:
```
root@ubuntu:/# http vault-vault:8200/v1/secret/qa/test "X-Vault-Token: X"
HTTP/1.1 200 OK
Cache-Control: no-store
Content-Length: 198
Content-Type: application/json
Date: Mon, 11 Dec 2017 19:43:08 GMT

{
    "auth": null, 
    "data": {
        "pass": "testpass", 
        "user": "teste"
    }, 
    "lease_duration": 2764800, 
    "lease_id": "", 
    "renewable": false, 
    "request_id": "1389fb75-9f31-baef-c22d-bc62df3e3579", 
    "warnings": null, 
    "wrap_info": null
}

root@ubuntu:/# http vault-vault:8200/v1/secret/qa/test "X-Vault-Token: X"
HTTP/1.1 200 OK
Cache-Control: no-store
Content-Length: 198
Content-Type: application/json
Date: Mon, 11 Dec 2017 19:43:19 GMT

{
    "auth": null, 
    "data": {
        "pass": "testpass", 
        "user": "teste"
    }, 
    "lease_duration": 2764800, 
    "lease_id": "", 
    "renewable": false, 
    "request_id": "5f1cdcbd-cd35-645b-312d-61dc9d532b0c", 
    "warnings": null, 
    "wrap_info": null
}

root@ubuntu:/# http vault-vault:8200/v1/secret/qa/test "X-Vault-Token: X"
HTTP/1.1 200 OK
Cache-Control: no-store
Content-Length: 198
Content-Type: application/json
Date: Mon, 11 Dec 2017 19:43:21 GMT

{
    "auth": null, 
    "data": {
        "pass": "testpass", 
        "user": "teste"
    }, 
    "lease_duration": 2764800, 
    "lease_id": "", 
    "renewable": false, 
    "request_id": "1fe759b8-7b52-5680-e5ec-9bcb6c5660c2", 
    "warnings": null, 
    "wrap_info": null
}
```

Request for each vault server individually.
```
root@ubuntu:/# http 10.60.1.46:8200/v1/secret/qa/test "X-Vault-Token: 1X"
HTTP/1.1 200 OK
Cache-Control: no-store
Content-Length: 198
Content-Type: application/json
Date: Mon, 11 Dec 2017 19:44:35 GMT

{
    "auth": null, 
    "data": {
        "pass": "testpass", 
        "user": "teste"
    }, 
    "lease_duration": 2764800, 
    "lease_id": "", 
    "renewable": false, 
    "request_id": "baefe869-0e39-bee3-394a-f81bcf9f87e6", 
    "warnings": null, 
    "wrap_info": null
}

root@ubuntu:/# http 10.60.2.52:8200/v1/secret/qa/test "X-Vault-Token: X"
HTTP/1.1 200 OK
Cache-Control: no-store
Content-Length: 198
Content-Type: application/json
Date: Mon, 11 Dec 2017 19:44:49 GMT

{
    "auth": null, 
    "data": {
        "pass": "testpass", 
        "user": "teste"
    }, 
    "lease_duration": 2764800, 
    "lease_id": "", 
    "renewable": false, 
    "request_id": "e833a73c-491c-8e02-2719-482add03de8f", 
    "warnings": null, 
    "wrap_info": null
}

root@ubuntu:/# http 10.60.2.53:8200/v1/secret/qa/test "X-Vault-Token: X"
HTTP/1.1 200 OK
Cache-Control: no-store
Content-Length: 198
Content-Type: application/json
Date: Mon, 11 Dec 2017 19:44:59 GMT

{
    "auth": null, 
    "data": {
        "pass": "testpass", 
        "user": "teste"
    }, 
    "lease_duration": 2764800, 
    "lease_id": "", 
    "renewable": false, 
    "request_id": "df61a83e-88c1-cedd-29e7-1e94ea6b663f", 
    "warnings": null, 
    "wrap_info": null
}
```
Getting the leader:
```
root@ubuntu:/# http 10.60.2.53:8200/v1/sys/leader                                                          
HTTP/1.1 200 OK
Cache-Control: no-store
Content-Length: 140
Content-Type: application/json
Date: Mon, 11 Dec 2017 19:45:59 GMT

{
    "ha_enabled": true, 
    "is_self": false, 
    "leader_address": "https://vault.hiden", 
    "leader_cluster_address": "https://10.60.2.52:8201"
}

root@ubuntu:/# http 10.60.1.46:8200/v1/sys/leader
HTTP/1.1 200 OK
Cache-Control: no-store
Content-Length: 140
Content-Type: application/json
Date: Mon, 11 Dec 2017 19:46:11 GMT

{
    "ha_enabled": true, 
    "is_self": false, 
    "leader_address": "https://vault.hiden", 
    "leader_cluster_address": "https://10.60.2.52:8201"
}

root@ubuntu:/# http 10.60.2.52:8200/v1/sys/leader
HTTP/1.1 200 OK
Cache-Control: no-store
Content-Length: 139
Content-Type: application/json
Date: Mon, 11 Dec 2017 19:46:18 GMT

{
    "ha_enabled": true, 
    "is_self": true, 
    "leader_address": "https://vault.hiden", 
    "leader_cluster_address": "https://10.60.2.52:8201"
}
```

Health:
```
root@ubuntu:/# http 10.60.2.52:8200/v1/sys/health
HTTP/1.1 200 OK
Cache-Control: no-store
Content-Length: 191
Content-Type: application/json
Date: Mon, 11 Dec 2017 19:47:10 GMT

{
    "cluster_id": "12d45e71-a6bf-7c96-d1cd-55ae1e3a50c5", 
    "cluster_name": "vault-cluster-85ba4038", 
    "initialized": true, 
    "sealed": false, 
    "server_time_utc": 1513021630, 
    "standby": false, 
    "version": "0.9.0"
}

root@ubuntu:/# http 10.60.2.53:8200/v1/sys/health
HTTP/1.1 429 Too Many Requests
Cache-Control: no-store
Content-Length: 190
Content-Type: application/json
Date: Mon, 11 Dec 2017 19:47:24 GMT

{
    "cluster_id": "12d45e71-a6bf-7c96-d1cd-55ae1e3a50c5", 
    "cluster_name": "vault-cluster-85ba4038", 
    "initialized": true, 
    "sealed": false, 
    "server_time_utc": 1513021644, 
    "standby": true, 
    "version": "0.9.0"
}

root@ubuntu:/# http 10.60.1.46:8200/v1/sys/health
HTTP/1.1 429 Too Many Requests
Cache-Control: no-store
Content-Length: 190
Content-Type: application/json
Date: Mon, 11 Dec 2017 19:47:29 GMT

{
    "cluster_id": "12d45e71-a6bf-7c96-d1cd-55ae1e3a50c5", 
    "cluster_name": "vault-cluster-85ba4038", 
    "initialized": true, 
    "sealed": false, 
    "server_time_utc": 1513021649, 
    "standby": true, 
    "version": "0.9.0"
}
```

Now 3 external requests:
```
$ http --verify=no "https://vault.hiden/v1/secret/qa/test" "X-Vault-Token: X" 
HTTP/1.1 200 OK
Cache-Control: no-store
Connection: keep-alive
Content-Length: 198
Content-Type: application/json
Date: Mon, 11 Dec 2017 19:47:57 GMT
Server: nginx/1.13.5
Strict-Transport-Security: max-age=15724800; includeSubDomains;

{
    "auth": null,
    "data": {
        "pass": "testpass",
        "user": "teste"
    },
    "lease_duration": 2764800,
    "lease_id": "",
    "renewable": false,
    "request_id": "1d8be04f-6401-b170-64c3-6f90e6c7c468",
    "warnings": null,
    "wrap_info": null
}


$ http --verify=no "https://vault.hiden/v1/secret/qa/test" "X-Vault-Token:X"
HTTP/1.1 200 OK
Cache-Control: no-store
Connection: keep-alive
Content-Length: 198
Content-Type: application/json
Date: Mon, 11 Dec 2017 19:47:59 GMT
Server: nginx/1.13.5
Strict-Transport-Security: max-age=15724800; includeSubDomains;

{
    "auth": null,
    "data": {
        "pass": "testpass",
        "user": "teste"
    },
    "lease_duration": 2764800,
    "lease_id": "",
    "renewable": false,
    "request_id": "edb468b1-c31f-a09f-9a0b-c53b543cf9fc",
    "warnings": null,
    "wrap_info": null
}


$ http --verify=no "https://vault.hiden/v1/secret/qa/test" "X-Vault-Token: X"
HTTP/1.1 200 OK
Cache-Control: no-store
Connection: keep-alive
Content-Length: 198
Content-Type: application/json
Date: Mon, 11 Dec 2017 19:48:00 GMT
Server: nginx/1.13.5
Strict-Transport-Security: max-age=15724800; includeSubDomains;

{
    "auth": null,
    "data": {
        "pass": "testpass",
        "user": "teste"
    },
    "lease_duration": 2764800,
    "lease_id": "",
    "renewable": false,
    "request_id": "b9037d2e-5c5d-56dd-27be-2a41a0d7bd0a",
    "warnings": null,
    "wrap_info": null
}
```

Do you want any more tests to be made? I can run then if necessary!

Sorry about some hiden stuff.

Reply:
Notes, I'm not 100% sure if the request that i've done will work in a non HA backend.
I"ve also added, but the templating in this is not good cause it relies on ingress to be configured. But i am not sure if this env have to be configured for my setup to work!
```
<           - name: VAULT_API_ADDR
<             value: "https://{{ index .Values.ingress.hosts 0 }}"
```